File: src/iq_gnc/src/square.cpp
#include <gnc_functions.hpp>
#include <iostream>
#include "std_msgs/String.h"
#include <cstring>
#include <fstream>
#include <string>

#include <sstream>

using namespace std;
//include API 

bool drone2Reach = false;
bool drone3Reach = false;	

string convertMessage(gnc_api_waypoint currentWayPoint) {
	std::string result;
	result += std::to_string(currentWayPoint.x) + ' ' + std::to_string(currentWayPoint.y) + ' ' + std::to_string(currentWayPoint.z) + ' ' + std::to_string(currentWayPoint.psi);
	return result;
}

void chatterCallback(const std_msgs::String::ConstPtr& msg) {
	drone2Reach = (msg->data == "drone 2 reached") ? true : false;
	drone3Reach = (msg->data == "drone 3 reached") ? true : false;
	if(drone2Reach && drone3Reach) {
		land();
	}
}

int main(int argc, char** argv)
{
	//initialize ros 
	ros::init(argc, argv, "drone1");
	ros::NodeHandle drone1("~");

	ros::Publisher chatter_pub = drone1.advertise<std_msgs::String>("controller", 1000);
	ros::Subscriber sub2 = drone1.subscribe("/drone2/square2/chatter2", 1000, chatterCallback);
	// ros::Subscriber sub3 = drone1.subscribe("/drone3/square3/chatter3", 1000, chatterCallback);
	
	//initialize control publisher/subscribers
	init_publisher_subscriber(drone1);

  	// wait for FCU connection
	wait4connect();

	//wait for used to switch to mode GUIDED
	set_mode("GUIDED");
	// wait4start();

	//create local reference frame 
	initialize_local_frame();

	//request takeoff
	takeoff(4);

	//specify some waypoints 
	// std::vector<gnc_api_waypoint> waypointList;
	// gnc_api_waypoint nextWayPoint;
	// nextWayPoint.x = 0;
	// nextWayPoint.y = 0;
	// nextWayPoint.z = 4;
	// nextWayPoint.psi = 0;
	// waypointList.push_back(nextWayPoint);
	// nextWayPoint.x = 5;
	// nextWayPoint.y = 0;
	// nextWayPoint.z = 4;
	// nextWayPoint.psi = -90;
	// waypointList.push_back(nextWayPoint);
	// nextWayPoint.x = 5;
	// nextWayPoint.y = 5;
	// nextWayPoint.z = 4;
	// nextWayPoint.psi = 0;
	// waypointList.push_back(nextWayPoint);
	// nextWayPoint.x = 0;
	// nextWayPoint.y = 5;
	// nextWayPoint.z = 4;
	// nextWayPoint.psi = 90;
	// waypointList.push_back(nextWayPoint);
	// nextWayPoint.x = 0;
	// nextWayPoint.y = 0;
	// nextWayPoint.z = 4;
	// nextWayPoint.psi = 180;
	// waypointList.push_back(nextWayPoint);
	// nextWayPoint.x = 0;
	// nextWayPoint.y = 0;
	// nextWayPoint.z = 0;
	// nextWayPoint.psi = 0;
	// waypointList.push_back(nextWayPoint);

	//specify control loop rate. We recommend a low frequency to not over load the FCU with messages. Too many messages will cause the drone to be sluggish
	ros::Rate rate(2.0);
	int numRead = 0;
	while(ros::ok())
	{
		std_msgs::String msg;

		string data;

		ros::spinOnce();
		rate.sleep();
		if(check_waypoint_reached() == 1) {
			if(numRead > 0) {
				cout << "ccc" << endl;
			} else {
				cout << "cc" << endl;
				ifstream inputFile("filename.txt");
				cout << "c" << endl;
				while(getline(inputFile, data)) {
					std::stringstream ss;
					ss << data;
					cout << data << endl;
					if(data.substr(0, 6) == "drone1") {
						numRead++;
						int drone2_msg_index = data.find(":");
						char drone2_content[100]; 
						strncpy(drone2_content, data.c_str() + 2 + drone2_msg_index, data.length() - 2 - drone2_msg_index);
						drone2_content[data.length() - 2 - drone2_msg_index] = '\0';
						// drone2_content[10] = '\0';
						float *arr_temp = new float[4];
						float arr[4] = {0};
						splitWayPointFromMsg(arr_temp, drone2_content);
						for(int i = 0; i < 4; i++) {
							arr[i] = *(arr_temp + i);
						}
						set_destination(arr[0], arr[1], arr[2], arr[3]);
					}

					msg.data = ss.str();

					ROS_INFO("%s", msg.data.c_str());
					chatter_pub.publish(msg);
				}
			}
		}
	}
	return 0;
}

File: src/iq_gnc/scripts/square.py
#! /usr/bin/env python
# Import ROS.
import rospy
# Import the API.
from iq_gnc.py_gnc_functions import *
# To print colours (optional).
from iq_gnc.PrintColours import *
from geometry_msgs.msg import Pose, PoseStamped, Point, Quaternion
from std_msgs.msg import String
from iq_gnc.ManageDrones import manageDrones, coordinates, LANDING, LANDED, TAKING_OFF, TAKEOFF

# Create an object for the API.
drone = gnc_api()

manageDrone = manageDrones(3)
        
def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.data)
    splitData = data.data.split(':')
    if splitData[0] == 'takeoff':
        manageDrone.updateReady(int(splitData[1][0:1]) - 2)  
    elif splitData[0] == 'reached':    
        manageDrone.updateSignal(int(splitData[1][0:1]) - 2)    
    
def handleDataFromServer(data):
    if manageDrone.state != LANDING:
        numDroneNeedToTakeOffArr = data.data.split()
        numDroneNeedToTakeOff = len(numDroneNeedToTakeOffArr)
        if manageDrone.state == LANDED:
            drone.takeoff(4)
            manageDrone.updateState('drone1 ', TAKING_OFF)
            manageDrone.needPublish = True
        elif manageDrone.checkEnough(numDroneNeedToTakeOff - 1) == False:
            manageDrone.needPublish = True
        manageDrone.requestXDroneTakeOff(numDroneNeedToTakeOff - 1)
        if(len(manageDrone.coordinatesAll) < 20):
            coordinates_str = data.data
            coordinates_arr = coordinates_str.split()   
            queue = []
            for pair in numDroneNeedToTakeOffArr:
                data = pair.split(',')
                newCoordinates = coordinates(int(data[0]) / 100, int("0"), int(data[1]) / 100)
                queue.insert(len(queue), newCoordinates)
            manageDrone.coordinatesAll.insert(0, queue)
        
def switch(topic, condition, countDownSleep, stay):
    if condition == 'landing':
        print('------ 0 Landing 0 ------')
        print('------ 1 You can\'t public your image 1 ------')
        if drone.check_waypoint_reached():
            manageDrone.updateState('drone1 ', LANDED)
    elif condition == 'takingoff':
        if manageDrone.needPublish:
            dataSend = 'takeoff: '
            for idx in range(len(manageDrone.requestTakeOff)):
                dataSend += 'drone' + str(idx + 2) + str(manageDrone.requestTakeOff[idx]) + ' '
            manageDrone.createData(dataSend)
            rospy.loginfo(manageDrone.dataSend)
            topic.publish(manageDrone.dataSend)
            manageDrone.needPublish = False
    elif condition == 'takeoff':
        if manageDrone.needPublish:
            dataSend = 'takeoff: '
            for idx in range(len(manageDrone.requestTakeOff)):
                action = 'land' if manageDrone.requestTakeOff[idx] == False else 'true' 
                dataSend += 'drone' + str(idx + 2) + str(action) + ' '
            manageDrone.createData(dataSend)
            rospy.loginfo(manageDrone.dataSend)
            topic.publish(manageDrone.dataSend)
            manageDrone.needPublish = False
        elif drone.check_waypoint_reached() and manageDrone.isReachAll():
            if countDownSleep > 0:
                if len(manageDrone.coordinatesAll) > 0:
                    manageDrone.unReachAll()
                    countDownSleep = 40
                    indexDestination = len(manageDrone.coordinatesAll) - 1
                    print('length: ' + str(indexDestination))
                    coor = manageDrone.coordinatesAll[indexDestination]
                    drone.set_destination(x=coor[0].coor_x, y=0, z=coor[0].coor_z, psi=coor[0].coor_psi)
                    for idx in range(1, len(coor)):
                        data = 'drone' + str(idx + 1) + ': ' + str(coor[idx].coor_x) + ' 0 ' + str(coor[idx].coor_z) + ' ' + str(coor[idx].coor_psi)
                        manageDrone.createData(data)
                        rospy.loginfo(manageDrone.dataSend)
                        topic.publish(manageDrone.dataSend)
                    manageDrone.coordinatesAll.pop(0)
                elif len(manageDrone.coordinatesAll) == 0:
                    countDownSleep -= 1
                print(str(countDownSleep))
            else:
                data = "land"
                rospy.loginfo(data)
                topic.publish(data)
                drone.land()
                manageDrone.updateState('drone1 ', LANDING)
                # rospy.loginfo(CGREEN2 + "All waypoints reached landing now." + CEND)
        elif stay % 4 == 0:
            # rospy.loginfo(manageDrone.dataSend)
            # topic.publish(manageDrone.dataSend)
            a = 1
        stay = (stay + 1) % 16
    elif condition == 'landed':
        print('---Landed---')
    return (countDownSleep, stay)
        
def main():
    # Initializing ROS node.
    rospy.init_node("drone1", anonymous=True)
    chatter_pub = rospy.Publisher(name='master', data_class=String, queue_size=100)
    
    # Subscribe for all sub drones
    rospy.Subscriber(name="/drone2/drone2_listen", data_class=String, queue_size=10, callback=callback)
    rospy.Subscriber(name="/drone3/drone3_listen", data_class=String, queue_size=10, callback=callback)
    rospy.Subscriber(name="/drone4/drone4_listen", data_class=String, queue_size=10, callback=callback)
    rospy.Subscriber(name="/server", data_class=String, queue_size=10, callback=handleDataFromServer)

    # Wait for FCU connection.
    drone.wait4connect()
    # Wait for the mode to be switched.
    # drone.wait4start()
    drone.set_mode("GUIDED")

    # Create local reference frame.
    drone.initialize_local_frame()
    # Request takeoff with an altitude of 3m.
    
    # Specify control loop rate. We recommend a low frequency to not over load the FCU with messages. Too many messages will cause the drone to be sluggish.
    rate = rospy.Rate(2)

    counter = 0
    stay = 0
    countDownSleep = 40
    while not rospy.is_shutdown():
        rate.sleep()
        print(manageDrone.state)
        if manageDrone.state == LANDED:
            counter = 0
        elif counter > 8:
            (countDownSleep, stay) = switch(chatter_pub, manageDrone.state, countDownSleep, stay)
            print(str(countDownSleep) + ' --- ' + str(stay))
        elif counter == 8:
            manageDrone.updateState('drone1 ', TAKEOFF)
            counter += 1
        elif manageDrone.state == TAKING_OFF:
            switch(chatter_pub, manageDrone.state, countDownSleep, stay)
            counter += 1
        else:
            print('waiting for signal')
            
            
if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        exit()
